src/secret/ledger_nano.rs

```diff
@@ -95,160 +101,214 @@ impl SecretManage for LedgerSecretManager {
+                return true;
+            }
+        }
+    }
+    // check if essence + bip32 indices fit into the buffer of the device
+    let essence_bytes = prepared_transaction.essence.pack_to_vec();
+    let total_size =
+        LedgerBIP32Index::default().packed_len() * prepared_transaction.inputs_data.len() + essence_bytes.len();
+
+    // return true if too large
+    total_size > buffer_size
+}
+
 #[async_trait]
 impl SecretManageExt for LedgerSecretManager {
     async fn sign_transaction_essence(&self, prepared_transaction: &PreparedTransactionData) -> crate::Result<Unlocks> {
         // lock the mutex to prevent multiple simultaneous requests to a ledger
         let _lock = self.mutex.lock().await;
 
-        // Get coin type and account index from first input
-        let (_coin_type, account_index) = match &prepared_transaction.inputs_data.first() {
-            Some(input) => {
-                match &input.chain {
-                    Some(chain) => {
-                        let raw: Vec<u32> = chain
-                            .segments()
-                            .iter()
-                            // XXX: "ser32(i)". RTFSC: [crypto::keys::slip10::Segment::from_u32()]
-                            .map(|seg| u32::from_be_bytes(seg.bs()))
-                            .collect();
-                        (raw[0], raw[1])
+        let mut input_bip32_indices: Vec<iota_ledger::LedgerBIP32Index> = Vec::new();
+        let mut coin_type: Option<u32> = None;
+        let mut account_index: Option<u32> = None;
+
+        let input_len = prepared_transaction.inputs_data.len();
+
+        for input in &prepared_transaction.inputs_data {
+            let bip32_indices: Vec<u32> = match &input.chain {
+                Some(chain) => {
+                    chain
+                        .segments()
+                        .iter()
+                        // XXX: "ser32(i)". RTFSC: [crypto::keys::slip10::Segment::from_u32()]
+                        .map(|seg| u32::from_be_bytes(seg.bs()))
+                        .collect()
+                }
+                None => return Err(crate::Error::NoInputs),
+            };
+            // coin_type and account_index should be the same in each output
+            assert!(
+                (coin_type.is_none() || coin_type == Some(bip32_indices[0]))
+                    && (account_index.is_none() || account_index == Some(bip32_indices[1]))
+            );
+            coin_type = Some(bip32_indices[1]);
+            account_index = Some(bip32_indices[2]);
+            input_bip32_indices.push(iota_ledger::LedgerBIP32Index {
+                bip32_change: bip32_indices[3] | HARDENED,
+                bip32_index: bip32_indices[4] | HARDENED,
+            });
+        }
+
+        assert!(coin_type.is_some() && account_index.is_some());
+
+        // unwrap values
+        let coin_type = coin_type.unwrap() & !HARDENED;
+        let bip32_account = account_index.unwrap() | HARDENED;
```

Comment:
Would be better to return errors here and not panic

Reply:
So I mean, it's not a case that actually ever could happen but if it does, something would be severly broken and therefore I wouldn't just return an error. But i wouldn't leave it completly unchecked ... assert is a bit of not ignoring^^

Reply:
hmm but if you like, we could change it, of course :innocent:

Reply:
Since when is it only in debug builds?
`Assertions are always checked in both debug and release builds, and cannot be disabled.` https://doc.rust-lang.org/std/macro.assert.html
And why can't it ever happen? I'm thinking about the offline example, where the inputs have the chain in the JSON, anyone could just provide something invalid there

Reply:
oh, interesting - I think it was C++ or something like this that disables the asserts checking in release builds and I remembered wrong :see_no_evil: :thinking: and thank you for your explanations, then I'll change it :ok_hand: